File: sdk/python/khp/state_bus.py
"""KHP State Bus — shared data layer for inter-device and agent communication."""

import asyncio
import time
from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional
from datetime import datetime, timezone


@dataclass
class Slot:
    """A named, typed data channel in the State Bus."""
    slot_id: str
    type: str = "float"
    unit: Optional[str] = None
    retention_s: int = 3600
    value: Any = None
    last_updated: Optional[str] = None
    history: List[dict] = field(default_factory=list)
    subscribers: List[str] = field(default_factory=list)
    max_history: int = 1000

    def write(self, value: Any):
        """Write a new value to this slot."""
        self.value = value
        self.last_updated = datetime.now(timezone.utc).isoformat()
        entry = {"value": value, "timestamp": self.last_updated}
        self.history.append(entry)
        if len(self.history) > self.max_history:
            self.history = self.history[-self.max_history:]

    def read(self) -> dict:
        """Read current value from this slot."""
        return {
            "slot_id": self.slot_id,
            "value": self.value,
            "type": self.type,
            "unit": self.unit,
            "last_updated": self.last_updated,
        }

    def get_history(self, last_n: int = 100) -> List[dict]:
        """Get recent history entries."""
        return self.history[-last_n:]


@dataclass
class Transform:
    """A composable data operation on slots."""
    transform_id: str
    input_slot: str
    operation: str
    params: dict = field(default_factory=dict)
    output_slot: Optional[str] = None
    output_event: Optional[str] = None

    def apply(self, value: Any) -> Any:
        """Apply this transform to a value."""
        if self.operation == "threshold":
            above = self.params.get("above")
            below = self.params.get("below")
            if above is not None and value > above:
                return {"triggered": True, "direction": "above", "value": value}
            if below is not None and value < below:
                return {"triggered": True, "direction": "below", "value": value}
            return {"triggered": False, "value": value}

        elif self.operation == "scale":
            factor = self.params.get("factor", 1.0)
            offset = self.params.get("offset", 0.0)
            return value * factor + offset

        elif self.operation == "moving_average":
            window = self.params.get("window", 10)
            return value  # Needs history context

        elif self.operation == "delta":
            return value  # Needs previous value

        elif self.operation == "clamp":
            min_v = self.params.get("min", float("-inf"))
            max_v = self.params.get("max", float("inf"))
            return max(min_v, min(max_v, value))

        return value


class StateBus:
    """Shared state bus — the communication backbone between devices and agents."""

    def __init__(self):
        self._slots: Dict[str, Slot] = {}
        self._transforms: Dict[str, Transform] = {}
        self._event_handlers: Dict[str, List[Callable]] = {}
        self._running = False

    def create_slot(self, slot_id: str, type: str = "float",
                    unit: str = None, retention_s: int = 3600) -> Slot:
        """Create a new data slot."""
        slot = Slot(slot_id=slot_id, type=type, unit=unit, retention_s=retention_s)
        self._slots[slot_id] = slot
        return slot

    def get_slot(self, slot_id: str) -> Optional[Slot]:
        """Get a slot by ID."""
        return self._slots.get(slot_id)

    def write_slot(self, slot_id: str, value: Any):
        """Write value to a slot and trigger any transforms."""
        slot = self._slots.get(slot_id)
        if slot is None:
            slot = self.create_slot(slot_id)
        slot.write(value)
        self._apply_transforms(slot_id, value)
        self._emit("slot_updated", {"slot_id": slot_id, "value": value})

    def read_slot(self, slot_id: str) -> Optional[dict]:
        """Read current value from a slot."""
        slot = self._slots.get(slot_id)
        if slot is None:
            return None
        return slot.read()

    def add_transform(self, transform: Transform):
        """Register a transform on the bus."""
        self._transforms[transform.transform_id] = transform

    def _apply_transforms(self, slot_id: str, value: Any):
        """Apply all transforms that watch this slot."""
        for t in self._transforms.values():
            if t.input_slot != slot_id:
                continue
            result = t.apply(value)
            if t.output_slot:
                self.write_slot(t.output_slot, result)
            if t.output_event:
                if isinstance(result, dict) and result.get("triggered"):
                    self._emit(t.output_event, result)

    def on(self, event: str, handler: Callable):
        """Subscribe to a bus event."""
        if event not in self._event_handlers:
            self._event_handlers[event] = []
        self._event_handlers[event].append(handler)

    def _emit(self, event: str, data: dict):
        """Emit an event to subscribers."""
        payload = {
            "event": event,
            "timestamp": datetime.now(timezone.utc).isoformat(),
            **data,
        }
        for handler in self._event_handlers.get(event, []):
            handler(payload)
        for handler in self._event_handlers.get("*", []):
            handler(payload)

    def list_slots(self) -> List[dict]:
        """List all active slots."""
        return [slot.read() for slot in self._slots.values()]

    def list_transforms(self) -> List[dict]:
        """List all registered transforms."""
        return [
            {
                "transform_id": t.transform_id,
                "input_slot": t.input_slot,
                "operation": t.operation,
                "output_slot": t.output_slot,
                "output_event": t.output_event,
            }
            for t in self._transforms.values()
        ]

File: sdk/typescript/src/state-bus.ts
/**
 * KHP State Bus — shared data layer for inter-device and agent communication.
 */

import type { SlotData, TransformDef, KHPEvent } from './types.js';

export class Slot {
  slotId: string;
  type: string;
  unit?: string;
  retentionS: number;
  value: unknown = null;
  lastUpdated?: string;
  history: Array<{ value: unknown; timestamp: string }> = [];
  subscribers: string[] = [];
  maxHistory: number;

  constructor(slotId: string, type = 'float', unit?: string, retentionS = 3600, maxHistory = 1000) {
    this.slotId = slotId;
    this.type = type;
    this.unit = unit;
    this.retentionS = retentionS;
    this.maxHistory = maxHistory;
  }

  write(value: unknown): void {
    this.value = value;
    this.lastUpdated = new Date().toISOString();
    this.history.push({ value, timestamp: this.lastUpdated });
    if (this.history.length > this.maxHistory) {
      this.history = this.history.slice(-this.maxHistory);
    }
  }

  read(): SlotData {
    return { slotId: this.slotId, value: this.value, type: this.type, unit: this.unit, lastUpdated: this.lastUpdated };
  }

  getHistory(lastN = 100): Array<{ value: unknown; timestamp: string }> {
    return this.history.slice(-lastN);
  }
}

export class Transform {
  transformId: string;
  inputSlot: string;
  operation: string;
  params: Record<string, unknown>;
  outputSlot?: string;
  outputEvent?: string;

  constructor(def: TransformDef) {
    this.transformId = def.transformId;
    this.inputSlot = def.inputSlot;
    this.operation = def.operation;
    this.params = def.params;
    this.outputSlot = def.outputSlot;
    this.outputEvent = def.outputEvent;
  }

  apply(value: unknown): unknown {
    if (this.operation === 'threshold') {
      const above = this.params.above as number | undefined;
      const below = this.params.below as number | undefined;
      const numValue = value as number;
      if (above !== undefined && numValue > above) return { triggered: true, direction: 'above', value };
      if (below !== undefined && numValue < below) return { triggered: true, direction: 'below', value };
      return { triggered: false, value };
    }
    if (this.operation === 'scale') {
      const factor = (this.params.factor as number) || 1.0;
      const offset = (this.params.offset as number) || 0.0;
      return (value as number) * factor + offset;
    }
    if (this.operation === 'clamp') {
      const min = (this.params.min as number) ?? -Infinity;
      const max = (this.params.max as number) ?? Infinity;
      return Math.max(min, Math.min(max, value as number));
    }
    return value;
  }
}

export class StateBus {
  private slots: Map<string, Slot> = new Map();
  private transforms: Map<string, Transform> = new Map();
  private eventHandlers: Map<string, Array<(event: KHPEvent) => void>> = new Map();

  createSlot(slotId: string, type = 'float', unit?: string, retentionS = 3600): Slot {
    const slot = new Slot(slotId, type, unit, retentionS);
    this.slots.set(slotId, slot);
    return slot;
  }

  getSlot(slotId: string): Slot | undefined {
    return this.slots.get(slotId);
  }

  writeSlot(slotId: string, value: unknown): void {
    let slot = this.slots.get(slotId);
    if (!slot) slot = this.createSlot(slotId);
    slot.write(value);
    this.applyTransforms(slotId, value);
    this.emit('slot_updated', { slotId, value });
  }

  readSlot(slotId: string): SlotData | null {
    const slot = this.slots.get(slotId);
    return slot ? slot.read() : null;
  }

  addTransform(def: TransformDef): void {
    this.transforms.set(def.transformId, new Transform(def));
  }

  on(event: string, handler: (event: KHPEvent) => void): void {
    if (!this.eventHandlers.has(event)) this.eventHandlers.set(event, []);
    this.eventHandlers.get(event)!.push(handler);
  }

  listSlots(): SlotData[] {
    return Array.from(this.slots.values()).map(s => s.read());
  }

  private applyTransforms(slotId: string, value: unknown): void {
    for (const t of this.transforms.values()) {
      if (t.inputSlot !== slotId) continue;
      const result = t.apply(value);
      if (t.outputSlot) this.writeSlot(t.outputSlot, result);
      if (t.outputEvent && typeof result === 'object' && result !== null && (result as any).triggered) {
        this.emit(t.outputEvent, result as Record<string, unknown>);
      }
    }
  }

  private emit(event: string, data: Record<string, unknown>): void {
    const payload: KHPEvent = { event, timestamp: new Date().toISOString(), ...data };
    for (const handler of this.eventHandlers.get(event) || []) handler(payload);
    for (const handler of this.eventHandlers.get('*') || []) handler(payload);
  }
}
